### PR TITLE
EDU-41: React Gen 2 support for adding editable regions to custom components

### PR DESCRIPTION
**Pull request by uttej-vsk**
## Description

Added React Gen 2 support for [Addition editable regions to Custom Components"](https://www.builder.io/c/docs/custom-components-children#adding-editable-regions-to-custom-components)

Once this PR is merged, the next goal is to update the code snippet in the docs.

**WIP** Jira ticket [EDU-41](https://builder-io.atlassian.net/browse/EDU-41)

[EDU-41]: https://builder-io.atlassian.net/browse/EDU-41?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by uttej-vsk** [comment]
@samijaber @sidmohanty11 Please review this PR. I will work on other frameworks for this ticket.

**Review by samijaber — CHANGES_REQUESTED** [review]
When adding support for a new snippet, it should always be a brand new component. We should not update an old component, or that means that we're editing an existing snippet. So `react/src/App.tsx` shouldn't be touched in this PR.

You are also missing tests for this.


The right approach would be to:
- add a `components/EditableRegions.tsx` (you pick the best name), whose content is the modified `App.tsx` in your PR

- import it in `react/src/main.tsx`:
```ts
const router = createBrowserRouter([
  {
    path: '/announcements/:id',
    element: <AnnouncementBar />,
  },
  {
    path: '/editable-regions', // feel free to pick another path name
    element: <EditableRegions />
  },
  {
    path: '/*',
    element: <App />,
  },
]);
```

- add playwright tests like in https://github.com/BuilderIO/builder/blob/main/packages/sdks-tests/src/snippet-tests/integrating-pages.spec.ts

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import { RegisteredComponent } from "../../../../output/react/types/context/types";
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:1]: this need not be a component.tsx file, I think we can name it as `registered-components.ts` or `custom-components.ts` or even `builder-custom-components.ts`

File: packages/sdks/snippets/react/src/App.tsx [hunk 2]
```diff
@@ -42,6 +42,7 @@ export default function App() {
       });
   }, []);
 
+
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/App.tsx:45]:
```suggestion
```

File: packages/sdks/snippets/react/src/App.tsx [hunk 3]
```diff
@@ -50,6 +51,6 @@ export default function App() {
 
   // return the page when found
   return (
-    <Content content={content} model={MODEL_NAME} apiKey={BUILDER_API_KEY} />
+    <Content content={content} model={MODEL_NAME} apiKey={BUILDER_API_KEY}/>
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/App.tsx:54]:
```suggestion
    <Content content={content} model={MODEL_NAME} apiKey={BUILDER_API_KEY} />
```

File: packages/sdks/snippets/react/src/components/AnnouncementBar.tsx [hunk 4]
```diff
@@ -36,6 +36,8 @@ export default function AnnouncementBar() {
       });
   }, []);
 
+  console.log(content)
+
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/AnnouncementBar.tsx:40]:
```suggestion
```

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
+/**
+ * https://www.builder.io/c/docs/custom-components-children
+ * src/components/customTabs.tsx
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:3]:
```suggestion
 * src/components/CustomTabs.tsx
```

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          {/* 
+            The tabList[] prop is an array that represents the tabs.
+            Each tab has a tabName, which is displayed as the button label,
+            and children, which are the content blocks rendered inside the tab.
+          */}
+          {props.tabList && props.tabList.map((tab, index) => (
+            <button
+              key={index}
+              className={activeTab === index ? 'active' : ''}
+              onClick={() => selectTab(index)}
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:42]:
```suggestion
              onClick={() => setActiveTab(index)}
```

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    id: string | undefined;
+  };
+}
+
+const CustomTabs = (props: TabProps) => {
+  const [activeTab, setActiveTab] = useState(0);
+
+  const selectTab = (tab: number) => {
+    setActiveTab(tab);
+  };
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:25]:
```suggestion
```

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  return (
+    <div>
+      <h2>Custom Component with editable regions</h2>
+
+      <div>
+        <div>
+          {/* 
+            The tabList[] prop is an array that represents the tabs.
+            Each tab has a tabName, which is displayed as the button label,
+            and children, which are the content blocks rendered inside the tab.
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:36]: should we mention that its coming from `inputs` declared inside `component-info` and link them to some docs that they can refer?

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+              onClick={() => selectTab(index)}
+            >
+              {tab.tabName}
+            </button>
+          ))}
+        </div>
+      </div>
+
+      {props.tabList && props.tabList.length !== 0 && (
+        <div>
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/components/CustomTabs.tsx:51]: i think for simplicity we can remove this and add optional chaining in the map function below?

File: packages/sdks/snippets/react/src/main.tsx [hunk 10]
```diff
@@ -3,16 +3,22 @@ import ReactDOM from 'react-dom/client';
...
   },
+  {
+    path: '/editable-region',
+    element: <EditableRegionRoute/>
+  },
   {
     path: '/*',
     element: <App />,
   },
+
```
Comment (sidmohanty11) [inline packages/sdks/snippets/react/src/main.tsx:21]:
```suggestion
```

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx (hunk 9 above)
Comment (uttej-vsk) [reply packages/sdks/snippets/react/src/components/CustomTabs.tsx:51]: Removed this and added optional chaining.

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx (hunk 8 above)
Comment (uttej-vsk) [reply packages/sdks/snippets/react/src/components/CustomTabs.tsx:36]: Done. Added a link

File: packages/sdks/snippets/react/src/components/CustomTabs.tsx (hunk 7 above)
Comment (uttej-vsk) [reply packages/sdks/snippets/react/src/components/CustomTabs.tsx:25]: Yeah, this is redundant. I have removed and updated this change.

**Review by samijaber — APPROVED** [review]

**Review by samijaber — COMMENTED** [review]
I would start by skipping the new tests for all the other frameworks that don't have them implemented, then do them separately.

**Pull request merged**